pkg/app/ops/insightcollector/collector.go

```diff
@@ -13,3 +13,356 @@
 // limitations under the License.
 
 package insightcollector
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"github.com/pipe-cd/pipe/pkg/filestore"
+	"github.com/pipe-cd/pipe/pkg/insightstore"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+var aggregateKinds = []model.InsightMetricsKind{
+	model.InsightMetricsKind_DEPLOYMENT_FREQUENCY,
+	model.InsightMetricsKind_CHANGE_FAILURE_RATE,
+}
+
+// InsightCollector implements the behaviors for the gRPC definitions of InsightCollector.
+type InsightCollector struct {
+	applicationStore datastore.ApplicationStore
+	deploymentStore  datastore.DeploymentStore
+	insightstore     insightstore.Store
+	logger           *zap.Logger
+}
+
+// NewInsightCollector creates a new InsightCollector instance.
+func NewInsightCollector(
+	ds datastore.DataStore,
+	fs filestore.Store,
+	logger *zap.Logger,
+) *InsightCollector {
+	a := &InsightCollector{
+		applicationStore: datastore.NewApplicationStore(ds),
+		deploymentStore:  datastore.NewDeploymentStore(ds),
+		insightstore:     insightstore.NewStore(fs),
+		logger:           logger.Named("insight-collector"),
+	}
+	return a
+}
+
+var (
+	pageSize = 50
+)
+
+func (i *InsightCollector) Run(ctx context.Context) error {
+	now := time.Now().UTC()
+	maxUpdateAt := now.Unix()
+
+	for {
+		//
+		// TODO: Modify to use page option instead of filter with updatedAt.
+		// related issue: [#1195](https://github.com/pipe-cd/pipe/issues/1195)
+		apps, err := i.applicationStore.ListApplications(ctx, datastore.ListOptions{
+			PageSize: pageSize,
+			Filters: []datastore.ListFilter{
+				{
+					Field:    "UpdatedAt",
```

Comment:
I think this should be done by `CreatedAt`.
Because `CreatedAt` is more stable than `UpdateAt`. The `UpdatedAt` value can be changed while querying and our targeting list could be re-ordered.
In our case, we just want a full list of applications, right?
The idea is sorting the list by `created_at` and use the `created_at` to query until getting all of them.

Reply:
Agree, fix it, thanks :D